Feature: raw sql

dsavina opened the pull request:
Added class `FindObjectsFromRawSqlQueryFactory` to handle object retrieving on full SQL queries.

It's particularly usefull on complex queries using `GROUP BY` and so on

For example:
```mysql
SELECT country.* AS ids
FROM country
JOIN users ON country.id= users.country_id
GROUP BY country.id
ORDER BY COUNT(users.id);
```

*Note: selecting the columns group `*` from the main table is mandatory!*

dsavina commented on tests/TDBMServiceTest.php:
```diff
@@ -225,6 +225,85 @@ public function testFindObjects()
+JOIN `users_roles` ON users.id = users_roles.user_id
+JOIN `roles` ON roles.id = users_roles.role_id
+GROUP BY users.id
+ORDER BY MAX(IF(roles.name = 'Admins', 3, IF(roles.name = 'Writers', 2, IF(roles.name = 'Singers', 1, 0)))) DESC
```
This will fail, as aggregate functions are not compatible with order direction in [thecodingmachine/magic-query](https://github.com/thecodingmachine/magic-query)... For now. This [pull request](https://github.com/thecodingmachine/magic-query/pull/37) will do.

moufmouf commented:
Just tagged magic-query v1.2.3.
We need to put this minimum version number in the composer.json, otherwise, unit tests testing the lowest dependencies number will fail.

Otherwise, very impressive PR! Good job!

dsavina commented:
Thanks! I've updated magic-query minimal version to v1.2.3. Also, I've added method findFromSql in generated DAOs, looking like:

```php
protected function findFromRawSql($sql, array $parameters = [], $countSql = null, $mode = null) : iterable
```

Without this last minor step, this new feature would not be of much use...

dsavina commented:
My bad, I'm mixing my Pull Requests... I've added this on package database.tdbm, for version v4.3. Coming right up on tdbm!

moufmouf commented:
Just add some documentation on the `findFromRawSql` DAO method in the /doc/quickstart.md file and I'm :+1: for merging!

dsavina commented:
Done!

moufmouf commented:
Yeah! Excellent!
